src/ray/gcs/gcs_client/service_based_accessor.cc

```diff
@@ -236,35 +239,43 @@ Status ServiceBasedActorInfoAccessor::AsyncSubscribeAll(
     const StatusCallback &done) {
   RAY_LOG(DEBUG) << "Subscribing register or update operations of actors.";
   RAY_CHECK(subscribe != nullptr);
-  subscribe_all_operation_ = [this, subscribe](const StatusCallback &done) {
-    auto on_subscribe = [subscribe](const std::string &id, const std::string &data) {
-      ActorTableData actor_data;
-      actor_data.ParseFromString(data);
-      subscribe(ActorID::FromBinary(actor_data.actor_id()), actor_data);
-    };
-    auto on_done = [this, subscribe, done](const Status &status) {
-      if (status.ok()) {
-        auto callback = [subscribe, done](
-                            const Status &status,
-                            const std::vector<rpc::ActorTableData> &actor_info_list) {
-          for (auto &actor_info : actor_info_list) {
-            subscribe(ActorID::FromBinary(actor_info.actor_id()), actor_info);
-          }
-          if (done) {
+  subscribe_all_operation_ =
+      [this, subscribe](const StatusCallback &done, bool is_pubsub_server_restarted) {
+        auto on_subscribe = [subscribe](const std::string &id, const std::string &data) {
+          ActorTableData actor_data;
+          actor_data.ParseFromString(data);
+          subscribe(ActorID::FromBinary(actor_data.actor_id()), actor_data);
+        };
+        auto on_done = [this, subscribe, done](const Status &status) {
+          if (status.ok()) {
+            auto callback =
+                [subscribe, done](
+                    const Status &status,
+                    const std::vector<rpc::ActorTableData> &actor_info_list) {
+                  for (auto &actor_info : actor_info_list) {
+                    subscribe(ActorID::FromBinary(actor_info.actor_id()), actor_info);
+                  }
+                  if (done) {
+                    done(status);
+                  }
+                  RAY_LOG(DEBUG)
+                      << "Finished subscribing register or update operations of actors.";
+                };
+            RAY_CHECK_OK(AsyncGetAll(callback));
+          } else if (done) {
             done(status);
           }
         };
-        RAY_CHECK_OK(AsyncGetAll(callback));
-      } else if (done) {
-        done(status);
-      }
-    };
-    auto status =
-        client_impl_->GetGcsPubSub().SubscribeAll(ACTOR_CHANNEL, on_subscribe, on_done);
-    RAY_LOG(DEBUG) << "Finished subscribing register or update operations of actors.";
-    return status;
-  };
-  return subscribe_all_operation_(done);
+
+        if (is_pubsub_server_restarted) {
+          return client_impl_->GetGcsPubSub().SubscribeAll(ACTOR_CHANNEL, on_subscribe,
+                                                           on_done);
+        }
+
+        on_done(Status::OK());
+        return Status::OK();
+      };
```

Comment:
The logic here is quite unclear, because:
1) when we call `AsyncSubscribeAll` for the first time, there is no `is_pubsub_server_restarted`. 
2) `on_done` seems to be called twice? Also the name `on_done` is ambiguous.

I think we can refactor it in this way:
1) save "subscrie_operation_` and `fetch_data_operation_` in different callbacks. 
2) when reconnected to GCS if `is_pubsub_server_restarted` is false, only call `fetch_data_operation`, otherwise call `subscrie_operation_`

Reply:
fixed